capa/features/extractors/binja/basicblock.py

```diff
@@ -0,0 +1,148 @@
+# Copyright (C) 2020 Mandiant, Inc. All Rights Reserved.
+# Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at: [package root]/LICENSE.txt
+# Unless required by applicable law or agreed to in writing, software distributed under the License
+#  is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and limitations under the License.
+
+import sys
+import string
+import struct
+from typing import Tuple, Iterator
+
+from binaryninja import Function, Variable
+from binaryninja import BasicBlock as BinjaBasicBlock
+from binaryninja import (
+    BinaryView,
+    VariableSourceType,
+    MediumLevelILSetVar,
+    MediumLevelILOperation,
+    MediumLevelILInstruction,
+)
+
+from capa.features.common import Feature, Characteristic
+from capa.features.address import Address, AbsoluteVirtualAddress
+from capa.features.basicblock import BasicBlock
+from capa.features.extractors.helpers import MIN_STACKSTRING_LEN
+from capa.features.extractors.binja.helpers import DisassemblyInstruction
+from capa.features.extractors.base_extractor import BBHandle, FunctionHandle
+
+
+def get_printable_len(il: MediumLevelILSetVar) -> int:
+    """Return string length if all operand bytes are ascii or utf16-le printable"""
+    width = il.dest.type.width
+    value = il.src.value.value
+
+    if width == 1:
+        chars = struct.pack("<B", value & 0xFF)
+    elif width == 2:
+        chars = struct.pack("<H", value & 0xFFFF)
+    elif width == 4:
+        chars = struct.pack("<I", value & 0xFFFFFFFF)
+    elif width == 8:
+        chars = struct.pack("<Q", value & 0xFFFFFFFFFFFFFFFF)
+    else:
+        return 0
+
+    def is_printable_ascii(chars_: bytes):
+        return all(c < 127 and chr(c) in string.printable for c in chars_)
+
+    def is_printable_utf16le(chars_: bytes):
+        if all(c == 0x00 for c in chars_[1::2]):
+            return is_printable_ascii(chars_[::2])
```

Comment:
I'm a big fan of local functions like this! they're great for readability while not polluting the global scope.

that being said, and more as a discussion point, is there a substantial performance impact of this pattern? in some languages, like javascript, this causes a new function object to be created for each invocation of the outer routine, which can cause a bunch of extra garbage for the GC. i'm not sure whether or not Python is smart enough to recognize that no captured variables are used and therefore a few function doesn't have to be created every iteration. do you?

Reply:
I am not good enough at Python to answer this. I will bring up this with the team and see if they have any ideas.

Reply:
Oh btw I think there is a problem with this `is_printable_ascii` function (which is copied from the IDA extractor). I compared with BN's logic of string detection, and I notice that our criteria is:

```C
static inline bool IsPrintableChar(uint8_t c)
{
	return (c == '\r') || (c == '\n') || (c == '\t') || ((c >= 0x20) && (c <= 0x7e));
}
```

Is this better? If so, should I change all instances of it?

Reply:
I found some interesting discussion at https://stackoverflow.com/questions/7839632/is-there-an-overhead-when-nesting-functions-in-python/7839697#7839697. It seems to me that there is some performance hit, but it should be fine.

Reply:
incidentally i saw this recent proposal to enhance list comprehensions in python, which naively compile to a local function object. the proposal removed the function object overhead and makes things slightly faster. in the microbenchmarks performance greatly improved, but only a small improvement on real world code. so, i think you're right, this isn't worth over-optimizing.

https://peps.python.org/pep-0709/